File: python/token_lib.py
# token_lib.py
# Shared helpers for manipulating XP token data.

import os
from typing import Any, Dict, List, Tuple, Optional

import bcrypt
from pymongo import MongoClient
from pymongo.collection import Collection
from pymongo.errors import DuplicateKeyError

MINUTE_BUCKETS = [15, 30, 45, 60]
CATEGORIES = ["regular", "weapon", "battlepass"]

MONGO_URI = os.environ.get("MONGODB_URI", "mongodb://localhost:27017")
MONGO_DB = os.environ.get("MONGODB_DATABASE", "codxp_tokens")
MONGO_COLLECTION = os.environ.get("MONGODB_USERS_COLLECTION", "users")

_USERS_COLLECTION: Optional[Collection] = None


def _get_users_collection() -> Collection:
    """Return the MongoDB collection storing user documents."""
    global _USERS_COLLECTION
    if _USERS_COLLECTION is None:
        client = MongoClient(MONGO_URI)
        collection = client[MONGO_DB][MONGO_COLLECTION]
        try:
            collection.create_index("username", unique=True)
        except Exception as exc:  # pragma: no cover - configuration error
            raise RuntimeError("Unable to initialise MongoDB user collection") from exc
        _USERS_COLLECTION = collection
    return _USERS_COLLECTION


def _default_tokens() -> Dict[str, List[int]]:
    return {cat: [0, 0, 0, 0] for cat in CATEGORIES}


def _default_profile() -> Dict[str, object]:
    return {"cod_username": "", "prestige": "", "level": 1}


def _to_int(value: Any) -> int:
    if isinstance(value, bool):
        return int(value)
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        return int(value)
    if isinstance(value, str):
        try:
            return int(value.strip())
        except ValueError:
            return 0
    return 0


def _ensure_size4(values: Any) -> List[int]:
    out = [0, 0, 0, 0]
    if isinstance(values, (list, tuple)):
        limit = min(4, len(values))
        for i in range(limit):
            out[i] = max(0, _to_int(values[i]))
    return out


def register_user(filename: str, username: str, password: str) -> bool:
    """Create a new user with a bcrypt-hashed password in MongoDB."""
    del filename  # Legacy argument retained for backwards compatibility
    collection = _get_users_collection()
    if collection.find_one({"username": username}):
        return False
    pw_hash = bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()
    doc: Dict[str, object] = {"username": username, "password_hash": pw_hash}
    doc.update(_default_profile())
    doc["tokens"] = _default_tokens()
    try:
        collection.insert_one(doc)
        return True
    except DuplicateKeyError:
        return False


def authenticate_user(filename: str, username: str, password: str) -> bool:
    """Validate supplied credentials against stored bcrypt hash."""
    del filename
    collection = _get_users_collection()
    doc = collection.find_one({"username": username})
    if not doc:
        return False
    hash_val = doc.get("password_hash")
    if not isinstance(hash_val, str) or not hash_val:
        return False
    return bcrypt.checkpw(password.encode(), hash_val.encode())


def read_all_tokens(filename: str, username: str) -> Dict[str, List[int]]:
    """Read tokens for the given user from MongoDB."""
    del filename
    collection = _get_users_collection()
    doc = collection.find_one({"username": username})
    if not doc:
        tokens = _default_tokens()
        collection.update_one(
            {"username": username},
            {
                "$set": {"tokens": tokens},
                "$setOnInsert": {"password_hash": "", **_default_profile()},
            },
            upsert=True,
        )
        return tokens
    tokens_obj = doc.get("tokens")
    data: Dict[str, List[int]] = {}
    for cat in CATEGORIES:
        raw_vals = []
        if isinstance(tokens_obj, dict):
            raw_vals = tokens_obj.get(cat, [])  # type: ignore[assignment]
        data[cat] = _ensure_size4(raw_vals)
    return data


def write_all_tokens(filename: str, username: str, data: Dict[str, List[int]]) -> None:
    """Write tokens for the given user back to MongoDB."""
    del filename
    collection = _get_users_collection()
    tokens: Dict[str, List[int]] = {}
    for cat in CATEGORIES:
        tokens[cat] = _ensure_size4(data.get(cat, []))
    collection.update_one(
        {"username": username},
        {
            "$set": {"tokens": tokens},
            "$setOnInsert": {"password_hash": "", **_default_profile()},
        },
        upsert=True,
    )


def compute_totals(tokens: List[int]) -> Tuple[int, float]:
    total_minutes = 0
    for count, minutes in zip(tokens, MINUTE_BUCKETS):
        total_minutes += count * minutes
    total_hours = total_minutes / 60.0
    return total_minutes, total_hours


def build_totals_report(data: Dict[str, List[int]]) -> str:
    lines: List[str] = []
    lines.append("=== 2XP Totals Report ===")
    grand_minutes = 0
    for cat in CATEGORIES:
        tokens = data[cat]
        cat_minutes, cat_hours = compute_totals(tokens)
        label = cat.capitalize()
        lines.append(f"{label}: {cat_minutes} minutes ({cat_hours:.2f} hours)")
        grand_minutes += cat_minutes
    lines.append("")
    lines.append(f"Grand Total: {grand_minutes} minutes ({grand_minutes/60.0:.2f} hours)")
    return "\n".join(lines)


def read_tokens_for_user(username: str) -> Dict[str, List[int]]:
    """Convenience wrapper mirroring the Java helper signature."""
    return read_all_tokens("", username)


def write_tokens_for_user(username: str, data: Dict[str, List[int]]) -> None:
    """Convenience wrapper mirroring the Java helper signature."""
    write_all_tokens("", username, data)

File: java/TokenLib.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.util.*;

public class TokenLib {
    public static final int[] MINUTE_BUCKETS = {15, 30, 45, 60};

    static void ensureFile(String filename) throws IOException {
        Path p = Paths.get(filename);
        if (!Files.exists(p)) {
            List<String> zeros = Collections.nCopies(12, "0");
            Files.write(p, zeros, StandardCharsets.UTF_8);
        }
    }

    private static List<Integer> parseInts(List<String> lines, int n) {
        List<Integer> out = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            try {
                out.add(Integer.parseInt(lines.get(i).trim()));
            } catch (Exception e) {
                out.add(0);
            }
        }
        return out;
    }

    public static Map<TokenCategory, List<Integer>> readAllTokens(String filename) throws IOException {
        ensureFile(filename);
        List<String> raw = Files.readAllLines(Paths.get(filename), StandardCharsets.UTF_8);
        Map<TokenCategory, List<Integer>> data = new EnumMap<>(TokenCategory.class);
        if (raw.size() >= 12) {
            data.put(TokenCategory.REGULAR, parseInts(raw.subList(0, 4), 4));
            data.put(TokenCategory.WEAPON, parseInts(raw.subList(4, 8), 4));
            data.put(TokenCategory.BATTLEPASS, parseInts(raw.subList(8, 12), 4));
        } else {
            data.put(TokenCategory.REGULAR, parseInts(raw, 4));
            data.put(TokenCategory.WEAPON, Arrays.asList(0, 0, 0, 0));
            data.put(TokenCategory.BATTLEPASS, Arrays.asList(0, 0, 0, 0));
        }
        return data;
    }

    private static List<Integer> ensureSize4(List<Integer> vals) {
        List<Integer> list = new ArrayList<>(Arrays.asList(0, 0, 0, 0));
        for (int i = 0; i < Math.min(4, vals.size()); i++) {
            list.set(i, vals.get(i));
        }
        return list;
    }

    public static void writeAllTokens(String filename, Map<TokenCategory, List<Integer>> data) throws IOException {
        List<String> out = new ArrayList<>();
        for (TokenCategory cat : TokenCategory.values()) {
            List<Integer> vals = ensureSize4(data.getOrDefault(cat, Arrays.asList(0, 0, 0, 0)));
            for (Integer v : vals) {
                out.add(String.valueOf(v));
            }
        }
        Files.write(Paths.get(filename), out, StandardCharsets.UTF_8);
    }

    public static AbstractMap.SimpleEntry<Integer, Double> computeTotals(List<Integer> tokens) {
        int total = 0;
        for (int i = 0; i < 4; i++) {
            int count = tokens.get(i);
            int minutes = MINUTE_BUCKETS[i];
            total += count * minutes;
        }
        double hours = total / 60.0;
        return new AbstractMap.SimpleEntry<>(total, hours);
    }

    public static String buildTotalsReport(Map<TokenCategory, List<Integer>> data) {
        List<String> lines = new ArrayList<>();
        lines.add("=== 2XP Totals Report ===");
        int grand = 0;
        for (TokenCategory cat : TokenCategory.values()) {
            List<Integer> tokens = data.get(cat);
            AbstractMap.SimpleEntry<Integer, Double> p = computeTotals(tokens);
            int catMinutes = p.getKey();
            double catHours = p.getValue();
            String label = cat.displayName();
            lines.add(label + ": " + catMinutes + " minutes (" + String.format(Locale.US, "%.2f", catHours) + " hours)");
            grand += catMinutes;
        }
        lines.add("");
        lines.add("Grand Total: " + grand + " minutes (" + String.format(Locale.US, "%.2f", grand / 60.0) + " hours)");
        return String.join("\n", lines);
    }
}

